ICU-21203 ICU-21198 ICU-21479 Improve docs discoverability for NumberFormatter

sffc opened the pull request:
<!--
Thank you for your pull request.
Please see http://site.icu-project.org/processes/contribute for general
information on contributing to ICU.

You will be automatically asked to sign the contributors license before the PR is accepted.
- sign: https://cla-assistant.io/unicode-org/icu
- license: http://www.unicode.org/copyright.html
-->

In this PR, I am addressing two docs-related issues related to NumberFormatter, which were opened by @filmil and @richgillam.  PTAL and let me know if these improvements look good to you.

##### Checklist

- [x] Issue filed:
    - https://unicode-org.atlassian.net/browse/ICU-21203
    - https://unicode-org.atlassian.net/browse/ICU-21198
    - https://unicode-org.atlassian.net/browse/ICU-21479
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [ ] Tests included
- [x] Documentation is changed or added

filmil reviewed (approved):
I of course don't get to allow you to merge things but I appreciate the changes. Having convenient links in the docs is most helpful!

Thank you for making the docs clearer!

richgillam reviewed (dismissed):
This does seem better than before; I can start in numberformatter.h and find my way to useful information, easier than before.  I think it's still somewhat hard to grok the whole thing if you're coming to it cold, but it's definitely improved.  It doesn't look like you've changed that much, though, so maybe it's partially that I know what I'm looking for now.  But the extra pointers you've added look good.

What I think I'd like to see the most is a new chapter in the user's guide, which I imagine is on your to-do list somewhere, but these changes look good to me.

richgillam reviewed (approved):
No complaints, but what does the `cnumtst.c` change have to do with docs discoverability?

sffc commented:
> No complaints

Thanks; squashed.

> but what does the `cnumtst.c` change have to do with docs discoverability?

I added it for https://unicode-org.atlassian.net/browse/ICU-21479, which involves making currency decimal formatting more discoverable, but I added the test to demonstrate that the claim made in the docs holds up.

sffc commented:
(I need an approval after the squash)
